src/tests/Microsoft.PowerFx.Interpreter.Tests/DeferredTypeTests.cs

```diff
@@ -0,0 +1,176 @@
+        [InlineData("Boolean(X)", "b")]
+        [InlineData("Index([1,2,3].Value, X)", "![Value:n]")]
+        [InlineData("Sum(X, 1) < 5", "b")]
+        [InlineData("Sum(X, 1, R) < 5", "b")] // All error are discarded for function calls, hence we don't get error for RecordType here.
+        [InlineData("Sum(X, T) < 5", "b")] // Since we discard all errors for function calls, Function calls are biased to non tabular overload.
+
+        [InlineData("Sum(X, X)", "n")]
+        [InlineData("X.Field1.Field1", "X")]
+        [InlineData("If(true, X)", "X")]
+        [InlineData("If(true, X, 1)", "n")]
+        [InlineData("If(true, X, X)", "X")]
+        [InlineData("ForAll([1,2,3], X)", "X")]
+        [InlineData("ForAll(ParseJSON(\"[1]\"), X)", "X")]
+        [InlineData("Abs(Table(X))", "n")]
+        [InlineData("Power(2, Table(X))", "n")]
+        [InlineData("Switch(X, 0, 0, 1, 1)", "n")]
+        [InlineData("Switch(0, 0, X, 1, 1)", "n")]
+        [InlineData("Switch(0, 0, X, 1, X)", "X")]
+        [InlineData("Switch(0, 0, X, 1, \"test\")", "s")]
+        [InlineData("Set(N, X); N", "n")]
+        [InlineData("Set(N, X); Set(N, 5); N", "n")]
+        [InlineData("Set(XM, X); XM", "X")]
+
+        // Ensures expression binds without any errors - but issues a warning for the deferred(unknown) type.
+        public void DeferredTypeTest(string script, string expectedReturnType)
+        {
+            var symbolTable = new SymbolTable();
+
+            symbolTable.AddVariable("X", FormulaType.Deferred);
+            symbolTable.AddVariable("R", RecordType.Empty());
+            symbolTable.AddVariable("T", RecordType.Empty().ToTable());
+            symbolTable.AddVariable("N", FormulaType.Number, mutable: true);
+            symbolTable.AddVariable("XM", FormulaType.Deferred, mutable: true);
+
+            TestDeferredTypeBindingWarning(script, Features.None, TestUtils.DT(expectedReturnType), symbolTable);
+        }
+
+        [Theory]
+        [InlineData("$\"Test {X} {R}\"", "Invalid argument type (Record). Expecting a Text value instead.")]
+        [InlineData("X + R", "Invalid argument type. Expecting one of the following: Number, Text, Boolean.")]
+        [InlineData("X.field + N.field", "Invalid use of '.'")]
+        [InlineData("Index([1,2,3], X).missing", "Name isn't valid. 'missing' isn't recognized")]
+        [InlineData("X < \"2021-12-09T20:28:52Z\"", "Invalid argument type. Expecting one of the following: Number, Date, Time, DateTime.")]
+        [InlineData("First(Sum(X, 1))", "Invalid argument type (Number). Expecting a Table value instead.")]
+
+        // Can't create aggregates around Deferred type.
+        [InlineData("[X]", "Incompatible type. The item you are trying to put into a table has a type that is not compatible with the table.")]
+        [InlineData("{test: X}", "Incompatible type. The item you are trying to put into a record has a type that is not compatible with the record.")]
```

Comment:
These are en-US error messages you're comparing against, it'll fail when @LucGenetier tries to run them. 

Generally, best practice for error tests is to compare the Message Key

Reply:
Great point! updated the tests.